### PR TITLE
Better 2D Curvilinear Operators (MATLAB)

### PR DESCRIPTION
**Pull request by bewagner1**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [X] Optimization
- [ ] Example
- [ ] Documentation

## Description

The current curvilinear operators have some issues.
- They use MATLAB interpolators, which are only 2nd order.
- They cannot handle periodic domains
- Their implementation does not follow what is outlined in the documentation very closely (nodal Jacobian)
- Their implementation is hard to follow (`DI2` and `GI2` functions)
- Their accuracy doesn't really change with _k_ and don't have _k_th order accuracy

This new implementation alleviates those issues, but still raises some of its own. These operators output to or act on the "extended" faces. These are the normal faces, plus the boundaries (see below).

Grad-Shafranov 4

## Related Issues & Documents

- [ ] I have created an Issue that is paired with this PR

## QA Instructions, Screenshots, Recordings

Run the file `curvConvergenceTest.m` in `examples/matlab_octave`. 

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [X] I need help with writing tests

## Read Contributing Guide and Code of Conduct

- [X] 📖 I have read the MOLE Contributing Guide: https://github.com/csrc-sdsu/mole/blob/main/CONTRIBUTING.md
- [X] 📖 I have read the MOLE Code of Conduct: https://github.com/csrc-sdsu/mole/blob/main/CODE_OF_CONDUCT.md

## [optional] Are there any post deployment tasks we need to perform?

Update documentation (re the "extended" faces)

### CODE REVIEW

**Comment by bewagner1** [comment]
Waiting on C++ periodic operators and C++ interpolators so that it can be implemented in C++ too.

**Comment by jbrzensk** [comment]
@bewagner1 This is still being looked at by @jcorbino. Thanks for your patience.

**Comment by jcorbino** [comment]
Great work, @bewagner1 🥇. The new architecture is much cleaner than the old interp2 + DI2/GI2 hybrid, the order of accuracy now genuinely scales with k, and routing periodic vs. non-periodic through dc and nc fits naturally with the rest of the library. The backward-compatible dispatch to *Legacy.m* is a thoughtful touch 👍🏻 

A few things I would like to see addressed before merging:

1. The boundary-row trick at the end of div2DCurv should be cleaned up.
`bdry = find(sum(spones(div2D(2,m,1,n,1,dc,nc)), 2) == 0);`
Building a full div2D with a hard-coded k = 2 only to read its empty-row pattern is wasteful, so your comment is spot on.

2. I think the convergence test never exercises the periodic branch. Will this be added later in a separate PR?

3. Minor housekeeping. The *Legacy.m* siblings need at least a deprecation note so we do not accumulate parallel implementations indefinitely. Also, grad3DCurv.m and div3DCurv.m have the same architectural issues this PR fixes in 2D. It would be worth opening a follow-up issue so the library does not sit in a mixed state for long.

This is a cool PR! Thanks for working on this!

**Comment by bewagner1** [comment]
Thanks for the feedback @jcorbino ! 

I added the updated 3D operators and cleaned up the boundary row finding scheme.

As for the periodic test in `curvConvergence.m`, the old operators don't do periodic boundary conditions so there's nothing really to compare. I could extend the domain of the problem to a full annulus but the old operators would have an extra 2 * (n + 2) boundary points to utilize, so it wouldn't be a one-to-one comparison (and in any realistic problem the value at those points is unknown).

**Comment by jcorbino** [comment]
Hi @bewagner1 - thanks for addressing those points.

What I meant by “I think the convergence test never exercises the periodic branch” wasn’t a comparison, but rather that the branch should be executed at least once to confirm it produces coherent results, no need to involve the old operators.

Let me know what you think. @jbrzensk your two cents on this?

I think the PR looks good. If you have a moment, please take one more pass through all the relevant files to ensure everything makes sense.

**Comment by bewagner1** [comment]
Ok I just made an example on a full annulus that is periodic in theta

**Comment by jcorbino** [comment]
Thanks @bewagner1 ! LGTM, please add `addpath('../../src/matlab_octave')` to the example.

https://github.com/csrc-sdsu/mole/issues/344 should be closed once this PR is merged.

**Comment by bewagner1** [comment]
Will do. Should I also modernize the `test_XYZCurv.m` examples?

**Comment by jcorbino** [comment]
> Will do. Should I also modernize the `test_XYZCurv.m` examples?

Yes please 👍🏻

**Comment by jcorbino** [comment]
Hi @bewagner1 — are you done working on this? I’d like to take one last look when it’s ready. No hurry, just let me know.
@jbrzensk feel free to jump in as well. So far, it LGTM.

**Comment by bewagner1** [comment]
> Hi @bewagner1 — are you done working on this? I’d like to take one last look when it’s ready. No hurry, just let me know. @jbrzensk feel free to jump in as well. So far, it LGTM.

Yes all done

File: src/matlab_octave/div3DCurv.m [hunk 1]
```diff
@@ -1,77 +1,126 @@
...
-    Dc = Div(:, m*(n-1)*(o-1)+(m-1)*n*(o-1)+1:end);
-    
-    % Apply transformation
-    Dx = J*(A*De+D*DI3(m-1, n-1, o-1, 'Dn')+G*DI3(m-1, n-1, o-1, 'Dc'));
-    Dy = J*(B*DI3(m-1, n-1, o-1, 'De')+E*Dn+H*DI3(m-1, n-1, o-1, 'Dcc'));
-    Dz = J*(C*DI3(m-1, n-1, o-1, 'Dee')+F*DI3(m-1, n-1, o-1, 'Dnn')+I*Dc);
-    
-    % Final 3D curvilinear mimetic divergence operator (d/dx + d/dy + d/dz)
+    if nargin ~= 4 && nargin ~= 12
+        error("div2DCurv:InvalidNumArgs", ...
```
Comment (jcorbino) [inline src/matlab_octave/div3DCurv.m]: div3DCurv, 4 or 12

File: src/matlab_octave/jacobian3D.m [hunk 2]
```diff
@@ -1,57 +1,100 @@
...
-    Yn = Y(mno+1:2*mno);
-    Yc = Y(2*mno+1:end);
-    Ze = Z(1:mno);
-    Zn = Z(mno+1:2*mno);
-    Zc = Z(2*mno+1:end);
-    
-    J = Xe.*(Yn.*Zc-Yc.*Zn)-Ye.*(Xn.*Zc-Xc.*Zn)+Ze.*(Xn.*Yc-Xc.*Yn);
+    if nargin ~= 4 && nargin ~= 12
+        error("jacobian3D:InvalidNumArgs", ...
+              "jacobian3D expects 4 or 9 arguments")
```
Comment (jcorbino) [inline src/matlab_octave/jacobian3D.m]: 4 or 12

File: src/matlab_octave/grad3DCurvLegacy.m [hunk 3]
```diff
@@ -0,0 +1,109 @@
+function G = grad3DCurvLegacy(k, X, Y, Z)
+% 
+% ----------------------------------------------------------------------------
+%                 !!! WARNING: DEPRICATED BY grad2DCurv.m !!!
```
Comment (jcorbino) [inline src/matlab_octave/grad3DCurvLegacy.m]: grad3DCurv.m

File: src/matlab_octave/div3DCurvLegacy.m [hunk 4]
```diff
@@ -0,0 +1,88 @@
+function D = div3DCurvLegacy(k, X, Y, Z)
+% 
+% ----------------------------------------------------------------------------
+%                 !!! WARNING: DEPRICATED BY div2DCurv.m !!!
```
Comment (jcorbino) [inline src/matlab_octave/div3DCurvLegacy.m]: div3DCurv.m

File: examples/matlab_octave/curvConvergenceTest.m [hunk 5]
```diff
@@ -0,0 +1,222 @@
+%% Convegence Test for Cuvilinear Operators
```
Comment (jcorbino) [inline examples/matlab_octave/curvConvergenceTest.m]: Convergence

File: src/matlab_octave/div3DCurv.m [hunk 6]
```diff
@@ -1,77 +1,126 @@
...
 %
 % DESCRIPTION
-%
+% Parameters:
+%                k : Order of accuracy
+%                X : x-coordinates (physical) of meshgrid centers if
+%                    optional arguments are specified, else nodes
+%                Y : y-coordinates (physical) of meshgrid centers if
+%                    optional arguments are specified, else nodes
+%                Z : z-coordintaes (physical) of meshgrid centers if
```
Comment (jcorbino) [inline src/matlab_octave/div3DCurv.m]: coordinates

File: src/matlab_octave/div3DCurv.m [hunk 7]
```diff
@@ -1,77 +1,126 @@
...
+%                    optional arguments are specified, else nodes
+%    (optional)  m : Number of cells in xi direction
+%    (optional) dx : Step size in xi direction
+%    (optional)  n : Number of cells in eta direction
+%    (optional) dy : Step size in eta direction
+%    (optional)  o : Number of cells in kappa direction
+%    (optional) dz : Step size in kappa direction
+%    (optional) dc : a0 (6x1 vector for left, right, bottom, top
```
Comment (jcorbino) [inline src/matlab_octave/div3DCurv.m]: front, and back.

File: src/matlab_octave/div3DCurv.m [hunk 8]
```diff
@@ -1,77 +1,126 @@
...
+%                    boundaries, resp.)
+%    (optional) nc : b0 (6x1 vector for left, right, bottom, top
```
Comment (jcorbino) [inline src/matlab_octave/div3DCurv.m]: front, and back.

File: src/matlab_octave/div2DCurvLegacy.m [hunk 9]
```diff
@@ -0,0 +1,70 @@
+function D = div2DCurvLegacy(k, X, Y)
+% 
+% ----------------------------------------------------------------------------
+%                 !!! WARNING: DEPRICATED BY div2DCurv.m !!!
```
Comment (jcorbino) [inline src/matlab_octave/div2DCurvLegacy.m]: DEPRICATED -> DEPRECATED - in all the *Legacy.m files.

**Review by jcorbino — CHANGES_REQUESTED** [review]
Hi @bewagner1 - I'm done with the review, please address the few comments I left. Everything else LGTM. Thanks!

**Review by jcorbino — APPROVED** [review]

**Comment by jbrzensk** [comment]
@bewagner1 Thanks Ben. Can you condense the commits down (git squash) to a few you think are important.

**Comment by jbrzensk** [comment]
How does the implementation work on curvilinear domains, ie where dx(dy,dz) is not fixed?

I gave the convergence_test.m a grid, the same grid, but with points clustered in one of the corners, and varying across the domain.
image

I get errorCur = 0.0012 and errorNew = 7.4761e+05. I get this because the nice thing is the old versions take the grid and figure everything else out, no need for spacing.

I do not know how to change/pick/give a varying dx and dy for this newer version.

**Comment by bewagner1** [comment]
Let me look into that.

The new operators use a computational domain of [0, 1] x [0, 1] x [0, 1] and dx should be 1 / m, dy should be 1 / n, and dz should be 1 / o

The old operators use a computational domain of [1, m] x [1, n] x [1, o] and have dx = dy = dz = 1

**Comment by bewagner1** [comment]
> The new operators use a computational domain of [0, 1] x [0, 1] x [0, 1] and dx should be 1 / m, dy should be 1 / n, and dz should be 1 / o

I misspoke. The new operators actually use a computational domain of [0, m * dx] x etc.

After looking into it a bit more @jbrzensk I do not know why the errors are so vastly different (in fact in my tests $k=2$ gave the best error). I would argue that the curvilinear operators are only as good as the mesh but my error is concentrated around $\theta = \frac{pi}{2}$ so idrk.

As for the choices for dx, dy, dz, they are for the computational domain and the spacing differences in the physical should get incorporated via the jacobian metrics.

**Comment by jcorbino** [comment]
@bewagner1 @jbrzensk just checking in on this, where did we land 🛩️ ?

File: examples/matlab_octave/curvConvergenceTest.m [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+errorCur = zeros(size(k));
+errorNew = zeros(size(k));
+
+dc = [1;1;1;1]; nc = [0;0;0;0];
+
+for i = 1:numel(k)
+
+    fprintf("Starting k = %d\n", k(i))
+    % Nodes to make current operators
+    rsN = 1:dy:2; tsN = pi:-dx:0;
```
Comment (jbrzensk) [inline examples/matlab_octave/curvConvergenceTest.m:79]: Need to test for cases where dx and dy are not fixed.

File: src/matlab_octave/grad2DCurv.m [hunk 11]
```diff
@@ -1,65 +1,85 @@
-function G = grad2DCurv(k, X, Y)
+function G = grad2DCurv(k, X, Y, m, dx, n, dy, dc, nc)
```
Comment (jbrzensk) [inline src/matlab_octave/grad2DCurv.m]: all grid variables should be extracted by the method. m and n can be calculated from X and Y. And dx and dy are not necessarily constant, so it seems unnecessary to put them as arguments, when the grad and div will use some uniform spacing, either 1 or 1/m.

File: src/matlab_octave/grad2DCurv.m [hunk 12]
```diff
@@ -1,65 +1,85 @@
...
+        In = speye(n+2);
+    end
+
+    % Make Ge and Gn act on and output to the centers
+    % This allows them to be added together
+    Ge = kron(In,IFCx) * kron(In,Ge);
+    Gn = kron(IFCy,Im) * kron(Gn,Im);
+
+    % Apply Metrics
+    [J,Xe,Xn,Ye,Yn] = jacobian2D(k,X,Y,m,dx,n,dy,dc,nc);
```
Comment (jbrzensk) [inline src/matlab_octave/grad2DCurv.m]: Same arguments for Jacobian, m,n are from X,Y, and dx,dy not constant, or if they are, set in jacobian2D.

**Review by jbrzensk — CHANGES_REQUESTED** [review]
I put this review as requesting changes, to trigger the review process again. dx,dy,dz are not constant in curvilinear space, so function calls, examples, and tests need to take this into account.

File: examples/matlab_octave/curvConvergenceTest.m (hunk 10 above)
Comment (bewagner1) [reply examples/matlab_octave/curvConvergenceTest.m:79]: The way dx and dy are specified in `curvConvergenceTest.m` are a slight misnomer. The are actually dr and d$\theta$. While these are fixed in `curvConvergenceTest.m`, meshes with undoubtedly not fixed dx and dy can be seen in all of the `test_XYZCurv` examples (excluding `test_div2DCurv.m`).

**Comment by bewagner1** [comment]
@jbrzensk Have you had a chance to review this yet? No worries if not

**Comment by jbrzensk** [comment]
@bewagner1 Sorry, will retest this today. So busy....

**Review by jbrzensk — APPROVED** [review]

File: examples/matlab_octave/curvConvergenceTest.m (hunk 10 above)
Comment (jbrzensk) [reply examples/matlab_octave/curvConvergenceTest.m:79]: Mispoke. Yes, technically they are not fixed, but, they are really close to each other, where errors could be masked by how close the cells are. This plot is the difference between rows, which implies it is uniform, even though they are not, just really close to each other (0.05).:
image

**Comment by jbrzensk** [comment]
For future sanity:
@bewagner1 fixed the spacing issues, which was the main issue when using a nonuniform, curvilinear, clustered grid.

I tested on clustered grid, with points cluster near the point where r=1.25, theta=pi/4.
image

While the 2nd order operator is still worse than the old one:
image

the new operators work for higher orders and exhibit the appropriate behavior for higher-order operators.
image

The initial error is not necessarily always worse, as there were situations where the new operators were ALWAYS better. This was exacerbated when the spacing along a boundary was not uniform, due to addScalarBC not supporting curvilinear boundaries (which require fixed dx and dy). Therefore, the plots of the convergence rate do not function when given these grids.
image
But, mixed with the addSCalarBC, I am not sure if calculating the errors on a curvilinear grid works the same for convergence rates without taking other things into consideration, so I am not going to hassle it.

**Comment by jbrzensk** [comment]
Also, the clustering code for others to use:
function x = clusterCenter(a,b,N,c,beta)
%CLUSTERCENTER Generate a one-dimensional stretched grid clustered about a point.
%
%   X = CLUSTERCENTER(A,B,N,C,BETA) generates N+1 grid points on the
%   interval [A,B] using a smooth hyperbolic-sine stretching that clusters
%   points about the location C.
%
%   INPUTS:
%       a      - Left endpoint of the interval.
%       b      - Right endpoint of the interval.
%       N      - Number of intervals (returns N+1 grid points).
%       c      - Location about which the grid is clustered.
%       beta   - Controls the width of the clustered region.
%                Larger values produce tighter clustering around C.
%                Smaller values approach a more uniform distribution.
%                A value of zero gives a uniform distribution.
%
%   OUTPUTS:
%       x      - Column vector of N+1 monotonically increasing grid points.
%
%   EXAMPLE:
%       % Generate 101 points on [1,2] clustered about r = 1.5
%       r = clusterCenter(1,2,100,1.5,0.10);
%
%       % Generate 65 angular points clustered about pi/2
%       theta = clusterCenter(0,pi,64,pi/2,0.20);
%
%   NOTES:
%       The mapping is based on the inverse hyperbolic sine:
%
%           x = c + beta*sinh(eta)
%
%       where eta is uniformly spaced between
%
%           asinh((A-c)/beta) and asinh((B-c)/beta).
%
%       This produces a smooth, monotonic, nonuniform grid with the
%       highest point density near C while preserving the interval
%       endpoints A and B.
%
% BRZENSKI July 2026

u = linspace(0,1,N+1);

eta = linspace( ...
    asinh((a-c)/beta), ...
    asinh((b-c)/beta), ...
    N+1);

x = c + beta*sinh(eta);

end
